### PR TITLE
feat!: adds input range check to optimize VerifyLeafHashes and VerifyInclusion methods

### PR DESCRIPTION
**Pull request by staheri14**
This PR optimizes the `VerifyLeafHashes` and `VerifyInclusion` methods by ensuring that the size of the provided leaves or leaf hashes matches the proof range and, if not, making an early return in order to prevent unnecessary hashing operations that would otherwise occur.
It introduces a breaking change by altering the behaviour of `VerifyLeafHashes` and `VerifyInclusion`. Previously, verification would succeed even if the provided leaves or leaf hashes exceeded the proof range, as long as the proof was valid and the extra leaves were appended at the end. In the new implementation, the verification will return false or an error in such cases.

### CODE REVIEW

File: proof.go [hunk 1]
```diff
@@ -13,7 +13,10 @@ import (
 )
 
 // ErrFailedCompletenessCheck indicates that the verification of a namespace proof failed due to the lack of completeness property.
-var ErrFailedCompletenessCheck = errors.New("failed completeness check")
+var (
+	ErrFailedCompletenessCheck = errors.New("failed completeness check")
+	ErrWrongLeafHashesSize     = errors.New("wrong leafHashes size")
+)
```
Comment (rootulp) [inline proof.go:19]: Sorry can't provide suggestion but I think the previous Go doc should be moved

```go
var (
	// ErrFailedCompletenessCheck indicates that the verification of a namespace proof failed due to the lack of completeness property.
	ErrFailedCompletenessCheck = errors.New("failed completeness check")
	ErrWrongLeafHashesSize     = errors.New("wrong leafHashes size")
)

```

**Review by rootulp — APPROVED** [review]
LGTM

**Comment by staheri14** [comment]
[To the reviewers] The codecov [error](https://github.com/celestiaorg/nmt/actions/runs/9068204992/job/24914967372?pr=254#step:8:32) is not specific to this PR, I am investigating it.

File: proof.go (hunk 1 above)
Comment (staheri14) [reply proof.go]: Thanks fixed it https://github.com/celestiaorg/nmt/pull/253/commits/b94a414c3fdd9dad617c5ba893f3f9b979792f78

**Review by rootulp — DISMISSED** [review]

**Review by walldiss — APPROVED** [review]

**Comment by staheri14** [comment]
Waiting on #255 to merge

**Review by rootulp — APPROVED** [review]

**Review by evan-forbes — APPROVED** [review]

**Pull request merged**